### ISSUE TITLE
Federated TS needs to handle TAN (Time Advance Notice) properly.

### ISSUE THREAD

**Issue by hokeun** [lf-lang/reactor-ts#94]
When there are physical actions and the earliest event has a tag ahead of physical time, a federate needs to send TAN (Time Advance Notice) to RTI instead of sending NET (Next Event Tag).

However, this is not implemented in reactor-ts. And we want to fix this.

More about TAN vs. NET:
https://github.com/lf-lang/reactor-c/blob/main/core/federated/net_common.h#L359

[Added Later]
- LF PR: https://github.com/lf-lang/lingua-franca/pull/1275
- Reactor-ts PR: https://github.com/lf-lang/reactor-ts/pull/106

**Comment by hokeun** [comment]
Here's the logic to send TAN:
https://github.com/lf-lang/reactor-c/blob/main/core/federated/federate.c#L2619

This logic is inside of `_lf_send_next_event_tag` at: https://github.com/lf-lang/reactor-c/blob/main/core/federated/federate.c#L2521

This is called inside of `send_next_event_tag` at: https://github.com/lf-lang/reactor-c/blob/main/core/threaded/reactor_threaded.c#L551

Then called here: https://github.com/lf-lang/reactor-c/blob/main/core/threaded/reactor_threaded.c#L592
From here, it looks like the second parameter `wait_for_reply` is always `true` for centralized coordination.

@Soroosh129 Hi Soroush, do you think you could confirm if my assumption about `wait_for_reply` is true? Also could you please explain what `wait_for_reply` is for?

**Comment by hokeun** [comment]
The code for sending NET in reactor-ts is here:
https://github.com/lf-lang/reactor-ts/blob/master/src/core/federation.ts#L521

Maybe we need to change the name of this function from `sendRTINextEventTime` to `sendRTINextEventTag` as well.

**Comment by hokeun** [comment]
@ByeongGil-Jun signed up to work on this issue.

**Comment by hokeun** [comment]
I just realized that this task is actually way more complicated than I expected. This involves checking for outputs that depend on physical actions, which must be checked within the code generator, then, indicated in the generated code.

Here is how this check is performed in the C generator:
https://github.com/lf-lang/lingua-franca/blob/c4dfbd9cebdb9aaf249508360e0bac1ce545458b/org.lflang/src/org/lflang/generator/c/CGenerator.java#L1922

This includes the call to `findOutputsConnectedToPhysicalActions`: https://github.com/lf-lang/lingua-franca/blob/c4dfbd9cebdb9aaf249508360e0bac1ce545458b/org.lflang/src/org/lflang/federated/FederateInstance.java#L497

The minDelay value is saved to `_fed.min_delay_from_physical_action_to_federate_output` [here](https://github.com/lf-lang/lingua-franca/blob/c4dfbd9cebdb9aaf249508360e0bac1ce545458b/org.lflang/src/org/lflang/generator/c/CGenerator.java#L1944) in the C generator.

Later, this `min_delay_from_physical_action_to_federate_output` value is checked inside `_lf_bounded_NET`: https://github.com/lf-lang/reactor-c/blob/main/core/federated/federate.c#L2619

The value of `min_delay_from_physical_action_to_federate_output` is NEVER (= LLONG_MIN, a negative value) by default. This default value indicates that there's no such output depending on a physical action.

This value is set to non-negative value when there's an output depending on a physical action in the federate. An example LF program including such output is [DistributedLoopedPhysicalAction.lf](https://github.com/lf-lang/lingua-franca/blob/master/test/C/src/federated/DistributedLoopedPhysicalAction.lf) in the LF tests. In this example, the `min_delay_from_physical_action_to_federate_output` value is set to zero.

We will need to take the following steps to correctly implement handling of TAN:

- In reactor-ts runtime, we add an indicator (i.e., `min_delay_from_physical_action_to_federate_output`) in each federate.
- In the code generator, we check if there's any output depending on a physical action, and code-generate to set the value of the indicator, `min_delay_from_physical_action_to_federate_output`.
- In reactor-ts runtime, we add a logic to check the value of `min_delay_from_physical_action_to_federate_output` as well as the current physical time, to decide whether we need to send NET or TAN.

**Comment by Soroosh129** [comment]
Just an update: We removed the TAN message entirely [here](https://github.com/lf-lang/reactor-c/commit/44dc5a1c341733575412f06628f4941da590795b). The reason for this removal was that the RTI was already treating TAN message exactly like Next Event Tag (NET) messages. However, the federates were not treating them the same way. The federates generally always produce a Logical Time Complete (LTC) for a promised NET eventually (unless it's NEVER or FOREVER), but they don't produce LTC for a promised TAN. This caused a lot of corner cases in the RTI that was hard to account for.

The alternative (and I think better) solution that we implemented is that federates with physical actions that are connected to outputs should regularly (as controlled by `coordination-options: {advance-message-interval: 10 msec}`) create dummy events locally that will force the federate to advance its tag and send NET messages to the RTI, followed by proper LTC messages. This ensures that downstream federates can make progress.

**Comment by hokeun** [comment]
Thanks for letting me know. It's such a bad timing :( I just finished implementing TAN in reactor-ts....

**Comment by Soroosh129** [comment]
> Thanks for letting me know. It's such a bad timing :( I just finished implementing TAN in reactor-ts....

I don't think this is wasted effort. Whenever you are sending a TAN message now, you could replace that with a dummy event.
In the C runtime, it was really a one-line change. You still need the logic that decides when to send the TAN (but now instead it creates a dummy event).

**Comment by hokeun** [comment]
> I don't think this is wasted effort. Whenever you are sending a TAN message now, you could replace that with the creations of a dummy event.

Glad to hear that!
 
> In the C runtime, it was really a one-line change. You still need the logic that decides when to send the TAN (but now instead it creates a dummy event).

Could you please point me to the one-line change?

Also, is min_delay_from_physical_action_to_federate_output still used? 

Thank you, @Soroosh129 !

**Comment by hokeun** [comment]
As we discussed today, I will make the TS runtime ignore PTAG and will add the dummy events to trigger NET (Next Event Tag) messages to be sent to RTI as a first step to let the downstream federate advance its TAG when there is an upstream federate with an output triggered by a physical action. Thank you @Soroosh129 and @lhstrh !

**Comment by lhstrh** [comment]
I'm not sure we concluded that `PTAG` should be ignored. What I recorded from our conversation is that `PTAG` should trigger input control reactions. Output control reactions, on the other hand, should be triggered at _any_ tag that has a local event associated with it. @Soroosh129, could you summarize here what the dummy events are for? I realize my notes did not include a description of those...

**Comment by Soroosh129** [comment]
> PTAG should trigger input control reactions

Right, but input (and output) control reactions are not yet implemented for the TS target.

> @Soroosh129, could you summarize here what the dummy events are for?

Dummy events do two things in the context of federates with physical actions:

1- Force a federate with a physical action (that is connected to an output port) to periodically try to advance its tag and thus produce a NET (which will allow downstream federates to advance their tag, allowing progress)
2- Trigger output control reactions, which unblock downstream federates (that are granted a PTAG) by sending them port absent messages

**Comment by lhstrh** [comment]
Thanks, @Soroosh129! Where does the code to schedule dummy events come from? The code generator? If so, it would be great if you can point us to the code where this happens in the `CGenerator`.

**Comment by hokeun** [comment]
> Where does the code to schedule dummy events come from? The code generator? If so, it would be great if you can point us to the code where this happens in the CGenerator.

I have been working on this for TSGeneartor. Here is the draft PR:
https://github.com/lf-lang/lingua-franca/pull/1275/

You can check my comment here: https://github.com/lf-lang/reactor-ts/issues/94#issuecomment-1141304132

**Comment by Soroosh129** [comment]
> Where does the code to schedule dummy events come from? 

It happens in [federate.c](https://github.com/lf-lang/reactor-c/blob/44dc5a1c341733575412f06628f4941da590795b/core/federated/federate.c#L2654)
